File: tests/misc/StringUtilTest.php
<?php

declare(strict_types=1);

use PHPUnit\Framework\TestCase;
use PHPUnit\Framework\Attributes\DataProvider;
use misc\StringUtil;

final class StringUtilTest extends TestCase
{
	// filterTagArray()

	public function testFilterTagArrayBasic(): void
	{
		$result = StringUtil::filterTagArray(['Wood', 'Metal', 'Stone']);
		$this->assertEqualsCanonicalizing(['wood', 'metal', 'stone'], $result);
	}

	public function testFilterTagArrayTrimsAndLowercases(): void
	{
		$result = StringUtil::filterTagArray(['  BRICK  ', ' Tile ']);
		$this->assertEqualsCanonicalizing(['brick', 'tile'], $result);
	}

	public function testFilterTagArraySplitsBySpace(): void
	{
		$result = StringUtil::filterTagArray(['red brick', 'blue stone']);
		$this->assertEqualsCanonicalizing(['red', 'brick', 'blue', 'stone'], $result);
	}

	public function testFilterTagArrayRemovesSpecialCharacters(): void
	{
		$result = StringUtil::filterTagArray(['wood!@#', 'metal$%^']);
		$this->assertEqualsCanonicalizing(['wood', 'metal'], $result);
	}

	public function testFilterTagArrayRemovesDuplicates(): void
	{
		$result = StringUtil::filterTagArray(['wood', 'Wood', 'WOOD']);
		$this->assertEquals(['wood'], array_values($result));
	}

	public function testFilterTagArraySkipsEmptyElements(): void
	{
		$result = StringUtil::filterTagArray(['', '   ', '!!!']);
		$this->assertEmpty($result);
	}

	public function testFilterTagArrayHandlesEmptyInput(): void
	{
		$result = StringUtil::filterTagArray([]);
		$this->assertEmpty($result);
	}

	// onlySmallLetters()

	public function testOnlySmallLettersBasic(): void
	{
		$this->assertEquals('helloworld', StringUtil::onlySmallLetters('Hello World!'));
	}

	public function testOnlySmallLettersRemovesNumbers(): void
	{
		$this->assertEquals('abc', StringUtil::onlySmallLetters('a1b2c3'));
	}

	public function testOnlySmallLettersReturnsEmptyForNoLetters(): void
	{
		$this->assertEquals('', StringUtil::onlySmallLetters('12345!@#'));
	}

	public function testOnlySmallLettersPreservesLowercase(): void
	{
		$this->assertEquals('alreadylower', StringUtil::onlySmallLetters('alreadylower'));
	}

	// onlyNumbers()

	public function testOnlyNumbersBasic(): void
	{
		$this->assertEquals('123', StringUtil::onlyNumbers('abc123def'));
	}

	public function testOnlyNumbersReturnsEmptyForNoDigits(): void
	{
		$this->assertEquals('', StringUtil::onlyNumbers('abcdef'));
	}

	public function testOnlyNumbersPreservesAllDigits(): void
	{
		$this->assertEquals('9876543210', StringUtil::onlyNumbers('9876543210'));
	}

	public function testOnlyNumbersStripsSpacesAndSpecials(): void
	{
		$this->assertEquals('42', StringUtil::onlyNumbers(' 4 2 '));
	}

	// removeNewline()

	public function testRemoveNewlineLineFeed(): void
	{
		$this->assertEquals('hello world', StringUtil::removeNewline("hello\n world"));
	}

	public function testRemoveNewlineCarriageReturn(): void
	{
		$this->assertEquals('hello world', StringUtil::removeNewline("hello\r world"));
	}

	public function testRemoveNewlineCRLF(): void
	{
		$this->assertEquals('hello world', StringUtil::removeNewline("hello\n\r world"));
	}

	public function testRemoveNewlineNoNewlines(): void
	{
		$this->assertEquals('hello world', StringUtil::removeNewline('hello world'));
	}

	// explodeFilterTrim()

	public function testExplodeFilterTrimBasic(): void
	{
		$result = StringUtil::explodeFilterTrim(',', 'a, b, c');
		$this->assertEquals(['a', 'b', 'c'], array_values($result));
	}

	public function testExplodeFilterTrimRemovesEmptyEntries(): void
	{
		$result = StringUtil::explodeFilterTrim(',', 'a,,b,,,c');
		$this->assertEquals(['a', 'b', 'c'], array_values($result));
	}

	public function testExplodeFilterTrimTrimsWhitespace(): void
	{
		$result = StringUtil::explodeFilterTrim(',', '  alpha  ,  beta  ,  gamma  ');
		$this->assertEquals(['alpha', 'beta', 'gamma'], array_values($result));
	}

	public function testExplodeFilterTrimEmptyString(): void
	{
		$result = StringUtil::explodeFilterTrim(',', '');
		$this->assertEmpty($result);
	}

	public function testExplodeFilterTrimThrowsOnEmptySeparator(): void
	{
		$this->expectException(\InvalidArgumentException::class);
		StringUtil::explodeFilterTrim('', 'a,b,c');
	}

	public function testExplodeFilterTrimCustomSeparator(): void
	{
		$result = StringUtil::explodeFilterTrim('|', 'x | y | z');
		$this->assertEquals(['x', 'y', 'z'], array_values($result));
	}

	// addHttpParameters()

	public function testAddHttpParametersToUrl(): void
	{
		$result = StringUtil::addHttpParameters('https://example.com/path', ['key' => 'value']);
		$this->assertStringContainsString('key=value', $result);
		$this->assertStringStartsWith('https://example.com', $result);
	}

	public function testAddHttpParametersMergesWithExisting(): void
	{
		$result = StringUtil::addHttpParameters('https://example.com/path?existing=1', ['new' => '2']);
		$this->assertStringContainsString('existing=1', $result);
		$this->assertStringContainsString('new=2', $result);
	}

	public function testAddHttpParametersOverridesExisting(): void
	{
		$result = StringUtil::addHttpParameters('https://example.com/path?key=old', ['key' => 'new']);
		$this->assertStringContainsString('key=new', $result);
		$this->assertStringNotContainsString('key=old', $result);
	}

	public function testAddHttpParametersMultipleParams(): void
	{
		$result = StringUtil::addHttpParameters('https://example.com/', ['a' => '1', 'b' => '2']);
		$this->assertStringContainsString('a=1', $result);
		$this->assertStringContainsString('b=2', $result);
	}

	public function testAddHttpParametersHandlesUrlWithoutScheme(): void
	{
		// A URL without scheme still produces a result with the parameters
		$result = StringUtil::addHttpParameters('/path/only', ['key' => 'value']);
		$this->assertStringContainsString('key=value', $result);
		$this->assertStringContainsString('/path/only', $result);
	}
}

File: internal/scraper/stringutil_test.go
package scraper

import (
	"testing"
)

// --- FilterTagArray ---

func TestFilterTagArrayBasic(t *testing.T) {
	result := FilterTagArray([]string{"Wood", "Metal", "Stone"})
	assertContainsAll(t, result, []string{"wood", "metal", "stone"})
}

func TestFilterTagArrayTrimsAndLowercases(t *testing.T) {
	result := FilterTagArray([]string{"  BRICK  ", " Tile "})
	assertContainsAll(t, result, []string{"brick", "tile"})
}

func TestFilterTagArraySplitsBySpace(t *testing.T) {
	result := FilterTagArray([]string{"red brick", "blue stone"})
	assertContainsAll(t, result, []string{"red", "brick", "blue", "stone"})
}

func TestFilterTagArrayRemovesSpecialCharacters(t *testing.T) {
	result := FilterTagArray([]string{"wood!@#", "metal$%^"})
	assertContainsAll(t, result, []string{"wood", "metal"})
}

func TestFilterTagArrayRemovesDuplicates(t *testing.T) {
	result := FilterTagArray([]string{"wood", "Wood", "WOOD"})
	if len(result) != 1 || result[0] != "wood" {
		t.Errorf("expected [wood], got %v", result)
	}
}

func TestFilterTagArraySkipsEmptyElements(t *testing.T) {
	result := FilterTagArray([]string{"", "   ", "!!!"})
	if len(result) != 0 {
		t.Errorf("expected empty, got %v", result)
	}
}

func TestFilterTagArrayHandlesEmptyInput(t *testing.T) {
	result := FilterTagArray([]string{})
	if len(result) != 0 {
		t.Errorf("expected empty, got %v", result)
	}
}

// --- OnlySmallLetters ---

func TestOnlySmallLettersBasic(t *testing.T) {
	assertStr(t, OnlySmallLetters("Hello World!"), "helloworld")
}

func TestOnlySmallLettersRemovesNumbers(t *testing.T) {
	assertStr(t, OnlySmallLetters("a1b2c3"), "abc")
}

func TestOnlySmallLettersReturnsEmptyForNoLetters(t *testing.T) {
	assertStr(t, OnlySmallLetters("12345!@#"), "")
}

func TestOnlySmallLettersPreservesLowercase(t *testing.T) {
	assertStr(t, OnlySmallLetters("alreadylower"), "alreadylower")
}

// --- OnlyNumbers ---

func TestOnlyNumbersBasic(t *testing.T) {
	assertStr(t, OnlyNumbers("abc123def"), "123")
}

func TestOnlyNumbersReturnsEmptyForNoDigits(t *testing.T) {
	assertStr(t, OnlyNumbers("abcdef"), "")
}

func TestOnlyNumbersPreservesAllDigits(t *testing.T) {
	assertStr(t, OnlyNumbers("9876543210"), "9876543210")
}

func TestOnlyNumbersStripsSpacesAndSpecials(t *testing.T) {
	assertStr(t, OnlyNumbers(" 4 2 "), "42")
}

// --- RemoveNewlines ---

func TestRemoveNewlinesLineFeed(t *testing.T) {
	assertStr(t, RemoveNewlines("hello\n world"), "hello world")
}

func TestRemoveNewlinesCarriageReturn(t *testing.T) {
	assertStr(t, RemoveNewlines("hello\r world"), "hello world")
}

func TestRemoveNewlinesCRLF(t *testing.T) {
	// \r\n is replaced first, then individual \n and \r
	result := RemoveNewlines("hello\n\r world")
	assertStr(t, result, "hello world")
}

func TestRemoveNewlinesNoNewlines(t *testing.T) {
	assertStr(t, RemoveNewlines("hello world"), "hello world")
}

// --- ExplodeFilterTrim ---

func TestExplodeFilterTrimBasic(t *testing.T) {
	result := ExplodeFilterTrim(",", "a, b, c")
	assertStrSlice(t, result, []string{"a", "b", "c"})
}

func TestExplodeFilterTrimRemovesEmptyEntries(t *testing.T) {
	result := ExplodeFilterTrim(",", "a,,b,,,c")
	assertStrSlice(t, result, []string{"a", "b", "c"})
}

func TestExplodeFilterTrimTrimsWhitespace(t *testing.T) {
	result := ExplodeFilterTrim(",", "  alpha  ,  beta  ,  gamma  ")
	assertStrSlice(t, result, []string{"alpha", "beta", "gamma"})
}

func TestExplodeFilterTrimEmptyString(t *testing.T) {
	result := ExplodeFilterTrim(",", "")
	if len(result) != 0 {
		t.Errorf("expected empty, got %v", result)
	}
}

func TestExplodeFilterTrimEmptySeparator(t *testing.T) {
	result := ExplodeFilterTrim("", "a,b,c")
	if result != nil {
		t.Errorf("expected nil for empty separator, got %v", result)
	}
}

func TestExplodeFilterTrimCustomSeparator(t *testing.T) {
	result := ExplodeFilterTrim("|", "x | y | z")
	assertStrSlice(t, result, []string{"x", "y", "z"})
}

// --- AddHTTPParameters ---

func TestAddHTTPParametersToURL(t *testing.T) {
	result, err := AddHTTPParameters("https://example.com/path", map[string]string{"key": "value"})
	if err != nil {
		t.Fatal(err)
	}
	assertContains(t, result, "key=value")
	assertContains(t, result, "https://example.com")
}

func TestAddHTTPParametersMergesWithExisting(t *testing.T) {
	result, err := AddHTTPParameters("https://example.com/path?existing=1", map[string]string{"new": "2"})
	if err != nil {
		t.Fatal(err)
	}
	assertContains(t, result, "existing=1")
	assertContains(t, result, "new=2")
}

func TestAddHTTPParametersOverridesExisting(t *testing.T) {
	result, err := AddHTTPParameters("https://example.com/path?key=old", map[string]string{"key": "new"})
	if err != nil {
		t.Fatal(err)
	}
	assertContains(t, result, "key=new")
	assertNotContains(t, result, "key=old")
}

func TestAddHTTPParametersMultipleParams(t *testing.T) {
	result, err := AddHTTPParameters("https://example.com/", map[string]string{"a": "1", "b": "2"})
	if err != nil {
		t.Fatal(err)
	}
	assertContains(t, result, "a=1")
	assertContains(t, result, "b=2")
}

func TestAddHTTPParametersPathOnly(t *testing.T) {
	result, err := AddHTTPParameters("/path/only", map[string]string{"key": "value"})
	if err != nil {
		t.Fatal(err)
	}
	assertContains(t, result, "key=value")
	assertContains(t, result, "/path/only")
}

// --- Helpers ---

func assertStr(t *testing.T, got, want string) {
	t.Helper()
	if got != want {
		t.Errorf("got %q, want %q", got, want)
	}
}

func assertStrSlice(t *testing.T, got, want []string) {
	t.Helper()
	if len(got) != len(want) {
		t.Errorf("got %v (len %d), want %v (len %d)", got, len(got), want, len(want))
		return
	}
	for i := range want {
		if got[i] != want[i] {
			t.Errorf("index %d: got %q, want %q", i, got[i], want[i])
		}
	}
}

func assertContainsAll(t *testing.T, got []string, want []string) {
	t.Helper()
	m := make(map[string]bool)
	for _, s := range got {
		m[s] = true
	}
	for _, w := range want {
		if !m[w] {
			t.Errorf("expected %q in %v", w, got)
		}
	}
}

func assertContains(t *testing.T, s, substr string) {
	t.Helper()
	if !contains(s, substr) {
		t.Errorf("expected %q to contain %q", s, substr)
	}
}

func assertNotContains(t *testing.T, s, substr string) {
	t.Helper()
	if contains(s, substr) {
		t.Errorf("expected %q NOT to contain %q", s, substr)
	}
}

func contains(s, substr string) bool {
	return len(s) >= len(substr) && searchStr(s, substr)
}

func searchStr(s, sub string) bool {
	for i := 0; i <= len(s)-len(sub); i++ {
		if s[i:i+len(sub)] == sub {
			return true
		}
	}
	return false
}
